### PR TITLE
Basic Node Server support with file, dir tree, and DAO handlers

### PR DESCRIPTION
**Pull request by mdittmer**

### CODE REVIEW

**Comment by mdittmer** [comment]
@kgrgreer PTAL @ code. Tests to be added to this PR for review before merge.

File: src/foam/net/node/Handler.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/**
+ * @license
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
Comment (mdittmer) [inline src/foam/net/node/Handler.js:3]: It's 2017.

File: src/foam/net/node/RestDAOHandler.js [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        if ( target !== '' ) {
+          self.send404(req, res);
+          self.warn('Attempt to put() with path suffix');
+          return true;
+        }
+
+        self.getPayload_(req).then(function(o) {
+          return self.dao.put(o);
+        }).then(function(o) {
+          self.send(res, 201, self.jsonify_(o));
```
Comment (mdittmer) [inline src/foam/net/node/RestDAOHandler.js]: This should be 200, not 201.

**Review by mdittmer — COMMENTED** [review]
... and I already found things I have to change.

File: src/foam/net/node/Handler.js (hunk 1 above)
Comment (jacksonic) [reply src/foam/net/node/Handler.js:3]: I can confirm, it is indeed 2017

**Comment by mdittmer** [comment]
This PR now depends on landing #254

**Comment by mdittmer** [comment]
@jacksonic and/or @kgrgreer PTAL.

File: src/foam/net/node/Server.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+      var self = this;
+      return new Promise(function(resolve, reject) {
+        self.server.close(function() {
+          self.server = null;
+          resolve(null);
+        });
+      });
+    },
+    function addHandler(handler) {
+      // TODO(adamvy): Not wild about this design, consider a better model for handling
```
Comment (jacksonic) [inline src/foam/net/node/Server.js]: Is this TODO still relevant? Or does it need to move

File: src/foam/net/node/FileHandler.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+foam.CLASS({
+  package: 'foam.net.node',
+  name: 'FileHandler',
+  extends: 'foam.net.node.Handler',
+
```
Comment (kgrgreer) [inline src/foam/net/node/FileHandler.js:22]: Add documentation:

File: src/foam/net/node/Handler.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+foam.CLASS({
+  package: 'foam.net.node',
+  name: 'Handler',
+
+  documentation:
+  function() {/*
```
Comment (kgrgreer) [inline src/foam/net/node/Handler.js]: Safe to use ES6 comments for node-only code.

File: src/foam/net/node/RestDAOHandler.js [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        }).catch(send500);
+      } else {
+        self.send404(req, res);
+        self.warn('Method not supported: '  + req.method);
+      }
+
+      return true;
+    },
+    {
+      name: 'fo2o_',
```
Comment (kgrgreer) [inline src/foam/net/node/RestDAOHandler.js:191]: Why isn't this just called objectify() to be compatible with the rest of FOAM?

File: src/foam/net/node/Server.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+foam.CLASS({
+  package: 'foam.net.node',
+  name: 'Server',
+
```
Comment (kgrgreer) [inline src/foam/net/node/Server.js:21]: documentation:

File: src/foam/net/node/StaticFileHandler.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+foam.CLASS({
+  package: 'foam.net.node',
+  name: 'StaticFileHandler',
+  extends: 'foam.net.node.Handler',
+
```
Comment (kgrgreer) [inline src/foam/net/node/StaticFileHandler.js:22]: documentation:

File: src/foam/net/node/StaticFileHandler.js [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+      },
+      factory: function() { return process.cwd(); }
+    },
+    {
+      class: 'String',
+      name: 'urlPath',
+      documentation: 'URL path prefix. Stripped before searching "dir".'
+    },
+    {
+      name: 'mimeTypes',
```
Comment (kgrgreer) [inline src/foam/net/node/StaticFileHandler.js:66]: Isn't this duplicated from above? Could it be a trait or in a LIB? Is this mutated? Could it be a constant instead?

File: src/foam/net/node/RestDAOHandler.js (hunk 6 above)
Comment (mdittmer) [reply src/foam/net/node/RestDAOHandler.js:191]: The idea was consistent naming for the two `something2somethingElse_` private functions.

File: src/foam/net/node/Server.js (hunk 3 above)
Comment (mdittmer) [reply src/foam/net/node/Server.js]: No. In fact, I'm happy to let the server register a flat list of handlers and support handler decoration (orthogonal to the server's logic) for things like OAuth.

File: src/foam/net/node/StaticFileHandler.js (hunk 9 above)
Comment (mdittmer) [reply src/foam/net/node/StaticFileHandler.js:66]: It was a constant in FOAM1. I wanted to give handlers the opportunity to provide their own `mimeTypes` map.

File: src/foam/net/node/RestDAOHandler.js (hunk 6 above)
Comment (mdittmer) [reply src/foam/net/node/RestDAOHandler.js:191]: @kgrgreer sound sensible?

File: src/foam/net/node/StaticFileHandler.js (hunk 9 above)
Comment (mdittmer) [reply src/foam/net/node/StaticFileHandler.js:66]: @kgrgreer does making it possible to override seem sensible?

**Comment by mdittmer** [comment]
@jingt06 FYI

File: src/foam/net/node/RestDAOHandler.js (hunk 6 above)
Comment (kgrgreer) [reply src/foam/net/node/RestDAOHandler.js:191]: It does make sense, just pointing out that it's inconsistent. I'm fine either way.

File: src/foam/net/node/StaticFileHandler.js (hunk 9 above)
Comment (kgrgreer) [reply src/foam/net/node/StaticFileHandler.js:66]: Yes, but it could be made into a MimeTypes Property, which would avoid duplication.

Comment (mdittmer) [reply src/foam/net/node/StaticFileHandler.js:66]: I think I understand. We can discuss IRL and clean up as future work.

**Pull request merged**